### PR TITLE
Rewrite xbcrypt using ds_decrypt

### PR DESCRIPTION
**Pull request by VasilyNemkov**
Blueprint:
[[https://blueprints.launchpad.net/percona-xtrabackup/+spec/xbcrypt-ds-decrypt-2.3](https://blueprints.launchpad.net/percona-xtrabackup/+spec/xbcrypt-ds-decrypt-2.3)]

Using ds_encrypt and ds_decrypt in xbcrypt for encryption and decryption.
Minor fixes to the ds_encrypt and ds_decrypt.

### CODE REVIEW

**Comment by VasilyNemkov** [comment]
param-build: http://jenkins.percona.com/job/percona-xtrabackup-2.3-param/317/

File: storage/innobase/xtrabackup/src/ds_decrypt.c [hunk 1]
```diff
@@ -68,7 +68,8 @@ typedef struct {
 	size_t			buf_size;
 } ds_decrypt_file_t;
 
-int		ds_decrypt_encrypt_threads = 1;
+uint		ds_decrypt_encrypt_threads = 1;
+my_bool		ds_decrypt_modify_file_extension = TRUE;
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/ds_decrypt.c]: small formatting issue - please make these two lines aligned

File: storage/innobase/xtrabackup/src/ds_encrypt.c [hunk 2]
```diff
@@ -60,6 +60,7 @@ typedef struct {
 /* Encryption options */
 uint		ds_encrypt_encrypt_threads;
 ulonglong	ds_encrypt_encrypt_chunk_size;
+my_bool		ds_encrypt_modify_file_extension = TRUE;
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/ds_encrypt.c]: same formatting issue here as well

File: storage/innobase/xtrabackup/test/t/xbcrypt.sh [hunk 3]
```diff
@@ -9,16 +9,29 @@
 
 encrypt_algo="AES256"
 encrypt_key="percona_xtrabackup_is_awesome___"
+parent_dir=$TEST_VAR_ROOT
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/test/t/xbcrypt.sh:12]: Nice catch regarding use if unbind variable `topdir`!

File: storage/innobase/xtrabackup/src/xbcrypt.c [hunk 4]
```diff
@@ -97,12 +90,26 @@ static struct my_option my_long_options[] =
...
+	 &opt_encrypt_threads, &opt_encrypt_threads, 0,
+	 GET_UINT, OPT_ARG, 1, 1, UINT_MAX, 0, 0, 0},
+
 	{"verbose", 'v', "Display verbose status output.",
 	 &opt_verbose, &opt_verbose,
 	  0, GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},
 	{0, 0, 0, 0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0}
 };
 
+/* Need the following definitions to avoid linking with ds_*.o and their link
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/xbcrypt.c]: But we do link with ds_*?

File: storage/innobase/xtrabackup/src/ds_encrypt.c [hunk 5]
```diff
@@ -156,11 +158,17 @@ encrypt_open(ds_ctxt_t *ctxt, const char *path, MY_STAT *mystat)
 
 	crypt_file = (ds_encrypt_file_t *) (file + 1);
 
-	/* Append the .xbcrypt extension to the filename */
-	fn_format(new_name, path, "", ".xbcrypt", MYF(MY_APPEND_EXT));
-	crypt_file->dest_file = ds_open(dest_ctxt, new_name, mystat);
+	if (ds_encrypt_modify_file_extension) {
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/ds_encrypt.c]: Please add the comment what is this needed for.

File: storage/innobase/xtrabackup/src/ds_decrypt.c [hunk 6]
```diff
@@ -146,12 +149,17 @@ decrypt_open(ds_ctxt_t *ctxt, const char *path, MY_STAT *mystat)
 
 	crypt_file = (ds_decrypt_file_t *) (file + 1);
 
-	/* Remove the .xbcrypt extension from the filename */
-	strncpy(new_name, path, FN_REFLEN);
-	new_name[strlen(new_name) - 8] = 0;
-	crypt_file->dest_file = ds_open(dest_ctxt, new_name, mystat);
+	if (ds_decrypt_modify_file_extension) {
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/ds_decrypt.c]: Please add the comment what is this needed for.

**Review by gl-sergei — CHANGES_REQUESTED** [review]
Looks good, just few minor comments

File: storage/innobase/xtrabackup/src/xbcrypt.c (hunk 4 above)
Comment (VasilyNemkov) [reply storage/innobase/xtrabackup/src/xbcrypt.c]: Nope, in fact we build few files of datasink as source files for the xbcrypt, the same way it is used in xbstream.

**Comment by VasilyNemkov** [comment]
Fixed

File: storage/innobase/xtrabackup/src/ds_decrypt.c [hunk 7]
```diff
@@ -146,12 +149,26 @@ decrypt_open(ds_ctxt_t *ctxt, const char *path, MY_STAT *mystat)
...
-	strncpy(new_name, path, FN_REFLEN);
-	new_name[strlen(new_name) - 8] = 0;
-	crypt_file->dest_file = ds_open(dest_ctxt, new_name, mystat);
+	/* xtrabackup and xbstream rely on fact that extension is appended on
+	   encryption and removed on decryption.
+	   That works well with piping compression and excryption datasinks,
+	   hinting on how to access contents of the file when it is needed.
+	   However, xbcrypt (and its users) does not expect such magic,
+	   and extension is set manually by the user or caller script.
+	   Here implicit extension modification causes more trouble than good.
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/ds_decrypt.c:158]: +1

File: storage/innobase/xtrabackup/src/ds_encrypt.c [hunk 8]
```diff
@@ -156,11 +158,26 @@ encrypt_open(ds_ctxt_t *ctxt, const char *path, MY_STAT *mystat)
...
-	crypt_file->dest_file = ds_open(dest_ctxt, new_name, mystat);
+	/* xtrabackup and xbstream rely on fact that extension is appended on
+	   encryption and removed on decryption.
+	   That works well with piping compression and excryption datasinks,
+	   hinting on how to access contents of the file when it is needed.
+	   However, xbcrypt (and its users) does not expect such magic,
+	   and extension is set manually by the user or caller script.
+	   Here implicit extension modification causes more trouble than good.
+
+	   See also ds_decrypt_modify_file_extension . */
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/ds_encrypt.c:169]: +1

**Review by gl-sergei — APPROVED** [review]

**Pull request merged**